File: LogiOS/TrustarApp/Repo/Chat/ChatRoomRepository.swift
//
//  ChatRoomRepository.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/11/25.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import Combine
import GRDB

class ChatRoomRepository {
    private var api: ChatApi
    private var chatDB: ChatDb
    init(api: ChatApi, chatDB: ChatDb) {
        self.api = api
        self.chatDB = chatDB
    }
    
    func chatRoomWithUsers(roomId: String) -> AnyPublisher<ChatRoomWithUsers?, Error> {
        return chatDB.chatRoomDao!.selectRoom(roomId: roomId)
            .map { [self] r -> AnyPublisher<ChatRoomWithUsers?, Error> in
                if r == nil { return Just(nil).setFailureType(to: Error.self).eraseToAnyPublisher()}
                return chatDB.chatRoomUserDao!.userListByRoom(roomId: r!.id)
                    .map { u in
                        ChatRoomWithUsers(room: r!, users: u)
                    }
                .eraseToAnyPublisher()
            }
            .switchToLatest()
            .eraseToAnyPublisher()
    }
    
    func chatRoomWithUsers2(roomId: String)-> AnyPublisher<ChatRoomWithUsers2?, Error> {
        return chatDB.chatRoomDao!.selectRoom(roomId: roomId)
            .map { [self] r -> AnyPublisher<ChatRoomWithUsers2?, Error> in
                if r == nil { return Just(nil).setFailureType(to: Error.self).eraseToAnyPublisher()}
                return chatDB.chatRoomUserDao!.userListByRoomWithReadRow(roomId: r!.id)
                    .map { u in
                        ChatRoomWithUsers2(room: r!, users: u)
                    }
                .eraseToAnyPublisher()
            }
            .switchToLatest()
            .eraseToAnyPublisher()
    }
    
    func reloadRooms() -> AnyPublisher<StateResult, NetworkError> {
        return api.getData(apiKey: Current.Shared.loggedUser?.token ?? "")
            .map {[self] rooms -> AnyPublisher<StateResult, NetworkError> in
                var delRows: [String] = []
                
                do {
                    var rs: [ChatRoom] = []
                    var us: [ChatUser] = []
                    var cs: [ChatRoomUser] = []
                    let version = UUID().uuidString
                    
                    for r in rooms {
                        rs.append(ChatRoom(id: r.talkRoomId,
                                           name: r.talkRoomName,
                                           image: r.talkRoomImageUrl,
                                           unread: r.unreadCount ?? 0,
                                           userCount: r.userCount ?? 0,
                                           lastUpdated: r.lastUpdateDatetime ?? 0,
                                           notification: r.notificationFlag ?? false,
                                           version: version))
                        
                        if let users = r.users {
                            for u in users {
                                cs.append(ChatRoomUser(roomId: r.talkRoomId, userId: u.userId, readMessageRowId: r.readRowsByUser()[u.userId]))
                                us.append(ChatUser(id: u.userId, name: u.userName, mail: u.mailAddress, avatar: u.avatarImageUrl))
                            }
                        }
                    }
                    
                    try chatDB.instanceDb?.write {db in
                        for r in rs {
                            try r.save(db)
                        }
                        chatDB.chatRoomDao?.deleteNot(db: db, version: version)
                        
                        chatDB.chatUserDao?.deleteAll(db: db)
                        for u in us {
                            try u.save(db)
                        }

                        chatDB.chatRoomUserDao?.deleteAll(db: db)
                        for c in cs {
                            try c.save(db)
                        }
                        
                        var ids: [String] = []
                        for r in rooms {
                            if let lr = r.deletedMessageItems {
                                for m in lr {
                                    ids.append(m.messageId)
                                    delRows.append(m.messageRowId)
                                }
                            }
                        }
                        
                        if !delRows.isEmpty {
                            chatDB.chatMessageDao?.deleteById(db: db, ids: ids)
                        }
                    }
                } catch {
                    debugPrint(error)
                }
                
                if delRows.isEmpty {
                    return Just(StateResult(result: true, message: "")).setFailureType(to: NetworkError.self).eraseToAnyPublisher()
                } else {
                    return api.updateDelete(body: MessageItemRows(messageRowIds: delRows, userInfo: Current.Shared.userInfo()), apiKey: Current.Shared.loggedUser?.token ?? "")
                }
            }
            .switchToLatest()
            .replaceError(with: StateResult(result: false, message: "ReloadRooms: Error!"))
            .setFailureType(to: NetworkError.self)
            .eraseToAnyPublisher()
    }
    
    func sortedRoomList() -> AnyPublisher<[ChatRoomExt], Error> {
        return chatDB.chatRoomDao!.selectRoomList()
            .map { lst in
                lst.sorted(by: { t1, t2 in
                    t1.room.lastUpdated > t2.room.lastUpdated
                })
            }.eraseToAnyPublisher()
    }
    
    func leaveRoom(roomId: String) -> AnyPublisher<String?, NetworkError> {
        let setting = ChatUserSettings(talkRoomId: roomId, userId: Current.Shared.userId(), notificationFlag: nil, leavingFlag: true, userInfo: Current.Shared.userInfo())
        return api.updateUserSetting(body: setting, apiKey: Current.Shared.loggedUser?.token ?? "")
            .map({ stt -> AnyPublisher<String?, NetworkError> in
                if stt.result {
                    self.chatDB.chatRoomDao?.deleteByRoomId(roomId: roomId)
                }
                
                return Just(stt.message).setFailureType(to: NetworkError.self).eraseToAnyPublisher()})
            .switchToLatest()
            .eraseToAnyPublisher()
    }
    
    func notificationOnOff(roomId: String, on: Bool) -> AnyPublisher<String?, NetworkError> {
        let setting = ChatUserSettings(talkRoomId: roomId, userId: Current.Shared.userId(), notificationFlag: on, leavingFlag: nil, userInfo: Current.Shared.userInfo())
        
        return api.updateUserSetting(body: setting, apiKey: Current.Shared.loggedUser?.token ?? "")
            .map({ $0.message }).eraseToAnyPublisher()
    }
    
    func getHistoryUsers() -> AnyPublisher<[TalkUser], NetworkError> {
        return api.getHistories(apiKey: Current.Shared.loggedUser?.token ?? "")
    }
    
    func getAllUsers() -> AnyPublisher<[TalkUser], NetworkError> {
        return api.getAllUsers(includeInvalid: false, apiKey: Current.Shared.loggedUser?.token ?? "")
    }
    
    func unreadCount() -> AnyPublisher<Int, Error> {
        chatDB.chatRoomDao!.selectRoomList()
            .map({ l in
                l.map( {$0.room.unread}).reduce(0, +)
            }).eraseToAnyPublisher()
    }
    
    func addRoom(users: [UserLite]) -> AnyPublisher<StateResult, NetworkError> {
        return api.addRoom(body: RoomCreate(user: users, userInfo: Current.Shared.userInfo()), apiKey: Current.Shared.loggedUser?.token ?? "")
    }
    
    func editRoom(roomId: String, users: [UserLite]) -> AnyPublisher<StateResult, NetworkError> {
        return api.addMember(body: RoomAddMember(roomId: roomId, users: users, userInfo: Current.Shared.userInfo()), apiKey: Current.Shared.loggedUser?.token ?? "")
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/repo/chat/ChatRoomRepository.kt
package jp.co.toukei.log.trustar.repo.chat

import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Single
import jp.co.toukei.log.lib.inTransaction
import jp.co.toukei.log.lib.observeOnIO
import jp.co.toukei.log.lib.orElseNull
import jp.co.toukei.log.lib.subscribeOnIO
import jp.co.toukei.log.trustar.Current
import jp.co.toukei.log.trustar.chat.ChatRoomWithUsers
import jp.co.toukei.log.trustar.chat.ChatRoomWithUsers2
import jp.co.toukei.log.trustar.compose.ComposeData
import jp.co.toukei.log.trustar.db.chat.entity.ChatRoom
import jp.co.toukei.log.trustar.db.chat.entity.ChatRoomExt
import jp.co.toukei.log.trustar.db.chat.entity.ChatRoomUser
import jp.co.toukei.log.trustar.db.chat.entity.ChatUser
import jp.co.toukei.log.trustar.rest.model.TalkUser
import jp.co.toukei.log.trustar.rest.post.ChatUserSettings
import jp.co.toukei.log.trustar.rest.post.MessageItemRows
import jp.co.toukei.log.trustar.rest.post.RoomAddMember
import jp.co.toukei.log.trustar.rest.post.RoomCreate
import jp.co.toukei.log.trustar.toCompletable
import java.util.Optional
import java.util.UUID
import java.util.concurrent.TimeUnit

class ChatRoomRepository {

    private val api = Current.chatApi

    fun chatRoomWithUsers(roomId: String): Flowable<Optional<ChatRoomWithUsers>> {
        return Current.chatDB.getInstance().chatRoomDao().selectRoom(roomId)
            .switchMap {
                val r = it.orElseNull()
                if (r == null) Flowable.just(Optional.ofNullable<ChatRoomWithUsers>(null))
                else {
                    Current.chatDB.getInstance().chatRoomUserDao().userListByRoom(r.id)
                        .map { u -> Optional.ofNullable(ChatRoomWithUsers(r, u)) }
                }
            }
            .throttleWithTimeout(100, TimeUnit.MILLISECONDS)
            .subscribeOnIO()
    }

    fun chatRoomWithUsers2(roomId: String): Flowable<Optional<ChatRoomWithUsers2>> {
        return Current.chatDB.getInstance().chatRoomDao().selectRoom(roomId)
            .switchMap {
                val r = it.orElseNull()
                if (r == null) Flowable.just(Optional.ofNullable<ChatRoomWithUsers2>(null))
                else {
                    Current.chatDB.getInstance().chatRoomUserDao().userListByRoomWithReadRow(r.id)
                        .map { u -> Optional.ofNullable(ChatRoomWithUsers2(r, u)) }
                }
            }
            .throttleWithTimeout(100, TimeUnit.MILLISECONDS)
            .subscribeOnIO()
    }

    fun reloadRooms(): Completable {
        return Single.defer { api.getData() }
            .observeOnIO()
            .subscribeOnIO()
            .flatMapCompletable { rooms ->
                val delRows = ArrayList<String>()
                Current.chatDB.getInstance().inTransaction {
                    val rs = ArrayList<ChatRoom>(rooms.size)
                    val s = rooms.sumOf { it.users?.size ?: 0 }
                    val us = ArrayList<ChatUser>(s)
                    val cs = ArrayList<ChatRoomUser>(s)

                    val version = UUID.randomUUID().toString()
                    rooms.forEach { r ->
                        rs += ChatRoom(
                            r.talkRoomId,
                            r.talkRoomName,
                            r.talkRoomImageUrl,
                            r.unreadCount ?: 0,
                            r.userCount ?: 0,
                            r.lastUpdateDatetime ?: 0L,
                            r.notificationFlag ?: false,
                            version
                        )
                        r.users?.forEach { u ->
                            cs += ChatRoomUser(r.talkRoomId, u.userId, r.readRowsByUser[u.userId])
                            us += ChatUser(u.userId, u.userName, u.mailAddress, u.avatarImageUrl)
                        }
                    }
                    chatRoomDao().apply {
                        updateOrIgnore(rs)
                        insert(rs)
                        deleteNot(version)
                    }
                    chatUserDao().apply {
                        deleteAll()
                        insertOrReplace(us.distinctBy(ChatUser::id))
                    }
                    chatRoomUserDao().apply {
                        deleteAll()
                        insertOrReplace(cs)
                    }
                    val ids = ArrayList<String>()
                    rooms.forEach { r ->
                        r.deletedMessageItems?.forEach { m ->
                            ids.add(m.messageId)
                            delRows.add(m.messageRowId)
                        }
                    }
                    if (delRows.isNotEmpty()) {
                        Current.chatDB.getInstance().chatMessageDao().deleteById(ids)
                    }
                }
                if (delRows.isEmpty()) {
                    Completable.complete()
                } else {
                    api.updateDelete(MessageItemRows(delRows)).toCompletable()
                }
            }
    }

    fun sortedRoomList(): Flowable<List<ChatRoomExt>> {
        return Flowable.defer {
            Current.chatDB.getInstance().chatRoomDao()
                .selectRoomList()
                .map { it.sortedByDescending { it.room.lastUpdated } }
        }
    }

    fun leaveRoom(roomId: String, reload: Boolean): Completable {
        return Single
            .defer {
                val it = ChatUserSettings(roomId, Current.userId, null, true)
                api.updateUserSetting(it)
            }
            .toCompletable()
            .observeOnIO()
            .doOnComplete { Current.chatDB.getInstance().chatRoomDao().deleteByRoomId(roomId) }
            .run {
                if (reload) andThen(reloadRooms())
                else this
            }
    }

    fun notificationOnOff(roomId: String, on: Boolean, reload: Boolean): Completable {
        return Single
            .defer {
                val it = ChatUserSettings(roomId, Current.userId, on, null)
                api.updateUserSetting(it)
            }
            .toCompletable()
            .run {
                if (reload) andThen(reloadRooms())
                else this
            }
    }

    fun getHistoryUsers(): Single<List<TalkUser>> {
        return api.getHistories().subscribeOnIO()
    }

    fun getAllUsers(): Single<List<TalkUser>> {
        return api.getAllUsers(false).subscribeOnIO()
    }

    fun unreadCount(): Flowable<Int> {
        return Current.chatDB.getInstance().chatRoomDao().selectRoomList().map { l ->
            l.sumOf { it.room.unread }
        }
    }

    fun addRoom(users: List<ComposeData.SelectChatUser>): Completable {
        return api.addRoom(RoomCreate(users))
            .observeOnIO()
            .toCompletable()
    }

    fun editRoom(roomId: String, users: List<ComposeData.SelectChatUser>): Completable {
        return api.addMember(RoomAddMember(roomId, users))
            .observeOnIO()
            .toCompletable()
    }
}
